File: easy/first_bad_version.rb
# https://leetcode.com/problems/first-bad-version/

# initial brute-force solution which is too slow to even submit (times out)

def first_bad_version(n)
  lowest = n

  (n).downto(0).each do |i|
    lowest = i if i < lowest && is_bad_version(i)
    break if !is_bad_version(i)
  end
  lowest
end

# submitted solution after looking at others' solutions

def first_bad_version(n)
  l, r = 0, n

  while (l < r) do
    mid = (l + r) / 2

    if is_bad_version(mid)
      r = mid
    else
      l = mid + 1
    end

    return l if l == r && is_bad_version(l)
  end
end

File: easy/first_bad_version.js
// https://leetcode.com/problems/first-bad-version/

var solution = function(isBadVersion) {
  /**
   * @param {integer} n Total versions
   * @return {integer} The first bad version
   */
  return function(n) {
    let l = 0;
    let r = n;

    while (l < r) {
      let mid = Math.floor((l + r) / 2);

      if (isBadVersion(mid)) {
        r = mid;
      } else {
        l = mid + 1;
      }

      if (l === r && isBadVersion(l)) return l;
    }
  };
};